### PR TITLE
[4 / 5] Make approval-voting runnable on a worker thread

### PR DESCRIPTION
**Pull request by alexggh**
This is part of the work to further optimize the approval subsystems, if you want to understand the full context start with reading https://github.com/paritytech/polkadot-sdk/pull/4849#issue-2364261568,

# Description
This PR contain changes to make possible the run of single approval-voting instance on a worker thread, so that it can be instantiated by the approval-voting-parallel subsystem. 

This does not contain any functional changes it just decouples the subsystem from the subsystem Context and introduces  more specific trait dependencies for each function instead of all of them requiring a context. 

This change can be merged  independent of the followup PRs.

### CODE REVIEW

File: polkadot/node/core/approval-voting/src/lib.rs [hunk 1]
```diff
@@ -1318,6 +1367,61 @@ where
...
+		sync_oracle,
+		metrics,
+		clock,
+		spawner,
+	);
+	let backend = DbBackend::new(db.clone(), approval_voting.db_config);
+	let spawner = approval_voting.spawner.clone();
+	spawner.spawn_blocking(
+		"approval-voting-rewrite-db",
+		Some("approval-voting-rewrite-subsystem"),
```
Comment (ordian) [inline polkadot/node/core/approval-voting/src/lib.rs:1406]: why is it called rewrite db/subsystem? as opposed to just approval-voting-subsystem?

**Review by ordian — APPROVED** [review]
Looks reasonable

**Review by alindima — APPROVED** [review]

File: polkadot/node/core/approval-voting/src/lib.rs (hunk 1 above)
Comment (alexggh) [reply polkadot/node/core/approval-voting/src/lib.rs]: This function is actually called and instantiated by the approval-voting-parallel in subsequent PR's, this name was just a placeholder, but the right approach is to not hardcode this here and allow the caller to pass in any name they want, which I did here: https://github.com/paritytech/polkadot-sdk/pull/4846/commits/377f4e1bcd1434a5dc3ecaef2f554afa2e9f168c

**Review by AndreiEres — APPROVED** [review]
LGTM, great job!

File: polkadot/node/network/approval-distribution/src/lib.rs [hunk 2]
```diff
@@ -646,6 +670,41 @@ enum MessageSource {
...
+	CryptoCheckFailed(InvalidAssignment),
+	// The assignment did not claim any valid candidate.
+	NoClaimedCandidates,
+	// Claimed invalid candidate.
+	#[allow(dead_code)]
+	ClaimedInvalidCandidateIndex {
+		claimed_index: usize,
+		max_index: usize,
+	},
+	// The assignment claimes more candidates than the maximum allowed.
```
[the remark this answers is not in the document]
Comment (alexggh) [reply polkadot/node/network/approval-distribution/src/lib.rs]: same as here: https://github.com/paritytech/polkadot-sdk/pull/4846#discussion_r1750217390

**Pull request merged**